tendermint-rs/src/public_keys.rs

```diff
@@ -0,0 +1,57 @@
+//! Public keys used in Tendermint networks
+// TODO:: account keys
+
+use sha2::{Digest, Sha256};
+use signatory::ed25519;
+use std::fmt::{self, Display};
+use subtle_encoding::bech32;
+
+/// Validator signing keys used for authenticating consensus protocol messages
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+pub enum ConsensusKey {
+    /// Ed25519 consensus keys
+    Ed25519(ed25519::PublicKey),
+}
+
+impl Display for ConsensusKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ConsensusKey::Ed25519(ref pk) => {
+                bech32::encode("cosmosvalconspub", pk.as_bytes()).fmt(f)
+            }
+        }
+    }
+}
+
+impl From<ed25519::PublicKey> for ConsensusKey {
+    fn from(pk: ed25519::PublicKey) -> ConsensusKey {
+        ConsensusKey::Ed25519(pk)
+    }
+}
+
+/// Secret Connection signing keys
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+pub enum SecretConnectionKey {
+    /// Ed25519 Secret Connection keys
+    Ed25519(ed25519::PublicKey),
+}
+
+impl Display for SecretConnectionKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            SecretConnectionKey::Ed25519(ref pk) => {
+                for val in Sha256::digest(pk.as_bytes()).iter().take(20) {
+                    write!(f, "{:x}", val)?;
```

Comment:
This is bogus... what should these be now? Are they upper case (non-colon-delimited) hex?

Reply:
Looks like you are looking how an `Address` is stringified? (the pub-key on the go side doesn't implement the Stringer interface but its address does): Address is a type alias for common.HexBytes which indeed stringifies via `strings.ToUpper(hex.EncodeToString(bz))`. E.g. for the public-key `PubKeyEd25519{F7FEB0B5BA0760B2C58893E329475D1EA81781DD636E37144B6D599AD38AA825}` the address would be the following 20 bytes: `117C95C4FD7E636C38D303493302D2C271A39669`

Reply:
Looks like I actually need: `write!(f, "{:02X}", byte)?`. I have this passing locally now... with a test even!

Reply:
Awesome!